### PR TITLE
Seeds SIWE implementation

### PR DESCRIPTION
**Pull request by sebastinez**
This PR adds the possibility to sign into a seed node with an Ethereum wallet.

**Authentication Process:**
- It sends a POST request to create an `Unauthorized` session
- With the returned `nonce` it creates a SIWE message and requests the wallet holder to sign in.
- The signed message and the raw message gets sent through a PUT request to the seed node
- If the verification on the seed node passes it updates the session to a `Authorized` session and returns the `session_id` and the session metadata.

Once authenticated the session ids get stored in `localStorage` and the currently active sessions get showed in the corresponding seed pages and the app header.

**TODOs left:**
- [x] Remove sessions from `localStorage` and the `session` store if expiration time is in the past.
- [x] Improve UI of session dropdown in the app header

### CODE REVIEW

File: src/base/seeds/View.svelte [hunk 1]
```diff
@@ -106,10 +119,36 @@
...
+              <Address address={siweSession.address} {config} resolve />
+            </div>
+            <div class="mobile">
+              <Address address={siweSession.address} {config} compact resolve />
+            </div>
+          {:else}
+            <SiweConnect
+              {seed}
+              {config}
+              connection={siweSession ? Connection.Connected : Connection.Disconnected}
```
Comment (cloudhead) [inline src/base/seeds/View.svelte:137]: This will always be `false`

File: src/base/seeds/View.svelte [hunk 2]
```diff
@@ -106,10 +119,36 @@
...
+            />
+          {/if}
+        {:else}
+          <div class="subtle">Please sign in to your wallet.</div>
```
Comment (cloudhead) [inline src/base/seeds/View.svelte:141]: I assume this is what you mean?
```suggestion
          <div class="subtle">Please connect your wallet.</div>
```

File: src/session.ts [hunk 3]
```diff
@@ -293,6 +304,27 @@ export async function changeAccounts(address: string): Promise<void> {
   state.refreshBalance(config);
 }
 
+export function loadSeedSessions(): { [key: string]: SeedSession } {
+  const siweStorage = localStorage.getItem("siwe");
+
+  if (siweStorage) {
+    const siwe: { [key: string]: SeedSession } = JSON.parse(siweStorage);
+
+    // We only keep the sessions that are still valid and remove expired from localStorage
```
Comment (cloudhead) [inline src/session.ts:313]:
```suggestion
    // We only keep the sessions that are still valid, and remove expired ones from `localStorage`.
```

File: src/session.ts [hunk 4]
```diff
@@ -293,6 +304,27 @@ export async function changeAccounts(address: string): Promise<void> {
...
+    // For a session to be valid the expiration time has to be bigger or equal than the current Date.
```
Comment (cloudhead) [inline src/session.ts:314]:
```suggestion
    // For a session to be valid the expiration time has to be bigger or equal than the current time.
```

File: src/siwe.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  if (! config.signer) {
+    return null;
+  }
+
+  const result = await createUnauthorizedSession(seed.api);
+  const address = await config.signer.getAddress();
+  const message = createSiweMessage(seed, address, result.nonce, config);
+  const signature = await config.signer.signMessage(message);
+
+  const auth: { id: string; session: SeedSession } = await new Request(`sessions/${result.id}`, seed.api).put({ message, signature: removePrefix(signature) });
```
Comment (cloudhead) [inline src/siwe.ts:55]: Wrap this line?

File: src/siwe.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  await new Request(`sessions/${result.id}`, seed.api).get({});
```
[the remark this answers is not in the document]
Comment (sebastinez) [reply src/siwe.ts:57]: At some point I was only returning, the nonce when updating a session, so I had to fetch the session object.
But can be removed now.

File: src/siwe.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+export interface SeedSession {
+  domain: string;
+  address: string;
+  statement: string;
+  uri: string;
+  version: string;
+  chain_id: string;
+  nonce: string;
+  issued_at: number;
+  expiration_time: number;
```
Comment (cloudhead) [inline src/siwe.ts:18]: Why are these snake_case?

**Pull request merged**